Projects: more rigorous linter settings

chadoh opened the pull request:
Fixes #987

This updates the linter settings in `apps/projects` only to extend:

    eslint:recommended
    plugin:react/recommended
    plugin:jsx-a11y/recommended

We violated these rules in many places, all of which have been addressed.

Quazia commented on apps/projects/app/script.js:
```diff
@@ -4,17 +4,12 @@ import { first } from 'rxjs/operators'
 import { app, initStore } from './store'
 
-retryEvery(async retry => {
+retryEvery(async () => {
```
If we're removing retry here I have a feeling using the retryEvery function here doesn't even make sense/we're not really using it.

chadoh replied:
Yeah, that seemed weird to me! Should I get rid of `retryEvery` while here? What should I replace it with?

ottodevs replied:
I think a good approach to still have the `retryEvery` feature is to pass it as param to the child function, as we did in Allocations, I am not that sure about the real utility of this helper function, but every core aragon app uses it.

I would say keep it like Allocations app and check if some warnings about `localhost:4444 postMessage origin = null` get solved

chadoh commented on apps/projects/app/components/Content/FilterTile.js:
```diff
@@ -24,18 +24,17 @@ export default class FilterTile extends PureComponent {
           >
             {this.props.text}
-            <div
+            <Button
```
The a11y linter complains when you add `onClick` to anything other than `<a>` or `<button>`

chadoh commented:
> do you have an opinion on if any of [the `console` statements] should actually be removed?

My general thought is that `console.error` and `console.warn` are ok, if not just there for debugging. I think all the ones that are still there are there to help us debug errors, when actual errors get thrown.

I've added a commit that updates the rule to allow this, without needing to add `eslint-disable-line` to each of them.

Now the only line with `eslint-disable-line` is the one in `ErrorBoundary`. I think it makes sense to keep this as a `console.log` instead of `warn` or `error`. 

> keep [`retryEvery`] like Allocations

Thanks for the tip, @ottodevs! I've added a commit to align the Projects `script.js` with the approach used in Allocations.

Quazia commented on apps/projects/app/store/app.js:
```diff
@@ -1,2 +0,0 @@
-import Aragon from '@aragon/api'
-export const app = new Aragon()
```
I don't think this should actually be removed, it's used in at least one place.

chadoh replied:
omg, yes it is, sorry about that. Re-added.

Quazia commented:
> About the better shaping / typing, I acknowledge that this involves heavy work, so I am ok with doing this more progressively, but I think definitely is something we would want to have for a production version
Yeah I think this falls outside of the scope of linting.

chadoh commented on apps/projects/app/components/Panel/ReviewApplication/ReviewApplication.js:
```diff
@@ -105,7 +108,15 @@ class ReviewApplication extends React.Component {
           <UserLink>
-            <img src={applicant.avatar} style={{ width: '32px', height: '32px', marginRight: '10px' }} />
+            <img
+              alt=""
```
It's _no_ alt that causes a11y issues. The recommendation is to either set it to something meaningful, or intentionally set as blank if the image is decorative-only. I don't think it would help users of screen readers understand this page to have text here. Do you?

chadoh commented on apps/projects/package.json:
```diff
@@ -57,17 +57,16 @@
     "eslint-plugin-standard": "^4.0.0",
     "parcel-bundler": "1.12.3",
-    "prettier": "^1.17.1",
+    "prettier-eslint": "^9.0.0",
```
It's a dependency of `prettier-eslint`. I'll re-add it though, if you prefer. I'm not sure if that helps clarify things or not.

ottodevs replied:
I don't have a strong opinion about this... as long as it does not start complaining about missing deerDependencies

I checked the module and it is not a peerDependency, so one can safely replace the other 🔀

chadoh commented on apps/projects/app/components/Content/IssueDetail/Markdown.js:
```diff
@@ -11,7 +11,11 @@ const Link = ({ children, ...props }) => (
 
-const ListItem = ({ checked, children, ...props }) => {
+Link.propTypes = {
+  children: PropTypes.element.isRequired,
```
I wasn't sure about this, either. The recommended linter settings required all propTypes to be defined, so I decided to just roll with it, rather than override. Since you asked about it here, and discovered that [this is intentional](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md#rule-options): the official recommendation there is to add `children` to `propTypes`. It encourages using `PropTypes.node`, however, instead of `PropTypes.element`. I'll update this PR accordingly.

chadoh commented on apps/projects/app/components/Card/Project.js:
```diff
@@ -101,15 +99,30 @@ const Project = ({
 }
 
+Project.propTypes = {
+  repoId: PropTypes.oneOfType([ PropTypes.string, PropTypes.number ]).isRequired,
```
I think I got an error when I had just `PropTypes.string`. I'll double-check.

chadoh replied:
No error; simplified; thanks!

ottodevs commented on apps/projects/app/components/Shared/BountyContextMenu.js:
```diff
@@ -78,6 +78,7 @@ const Item = styled(ContextMenuItem)`
 
 BountyContextMenu.propTypes = {
   issue: PropTypes.shape({
+    requestsData: PropTypes.array.isRequired,
```
I was wondering about this one, I think there is a big chance that there takes a while between the prop name declaration, and when the component receives its value, so I would say removing the requirement should be safe enough to fix the warning, as we did on other components, I didn't see any other apparent side-effects when the prop is `undefined`
I let the final decision to your criteria on that 👌

chadoh replied:
I updated it, and correspondingly updated the code to check if it's there before assuming it has a `length` property

chadoh commented:
@ottodevs 

> better define and shape lot of PropTypes... even reusing some PropTypes for common places, things like issues or projects

I like this idea, but I would rather accomplish it with TypeScript.

> design regression with the filtered repo label

This was because of replacing a div with a Button, and not improving the styling. I've added a commit to fix it. Thanks for the catch!

> lot of "unselectable" components

Looks like I only removed two `unselectable` imports. One of them seems to have previously been used with a component that wraps an aragonUI `<Button>`, which already adds `user-select: none` ([check it](https://ui.aragon.org/button/)). The other spot is the filter buttons. Though these do not have `user-select: none`, I was not able to select the text in them, in either Brave or Firefox. 

> received this message while trying to fund issue

I see the same error on `dev` ✅ 

---

I think it's ready for another review, @Quazia @ottodevs
